# Data Cleaning Project 2:

In [1]:
# Importing libraries 

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Loading the dataset

df = pd.read_csv("hotel-booking-data.txt", sep="\t")
df.head()

Date Company        Person Name  Room number
0  1-Jan-2022  Avamba  Anatole Ridehalgh       4008.0
1  1-Jan-2022    Fatz    Aldrich McKevin       2002.0
2  1-Jan-2022   Leexo    Stanley Hadrill       4012.0
3      Hotels     NaN                NaN          NaN
4  1-Jan-2022  Rhyzio       Lyndell Tice       1006.0

In [3]:
# Column names

df.columns

Index(['Date', 'Company', 'Person Name', 'Room number'], dtype='object')

In [4]:
# Information of the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         195 non-null    object 
 1   Company      134 non-null    object 
 2   Person Name  153 non-null    object 
 3   Room number  153 non-null    float64
dtypes: float64(1), object(3)
memory usage: 6.2+ KB


In [5]:
# Missing Values

df.isnull().sum()

Date            0
Company        61
Person Name    42
Room number    42
dtype: int64

In [14]:
# Summary Statistics
df.describe()

Room number
count   134.000000
mean   2947.843284
std    1664.805780
min    1001.000000
25%    2001.000000
50%    3006.500000
75%    4007.750000
max    7002.000000

- As it is a categorical data we cannot get the proper summary statistics for the dataset.
- Hence, the **room number column** has float data type, its only generating statistics regarding that column.

In [15]:
# Value count of companies by booking

df["Company"].value_counts()

Leexo           3
Fivechat        3
Quinu           3
Jabbersphere    2
Twitterlist     2
               ..
Topiclounge     1
Shufflebeat     1
Skyvu           1
Youspan         1
Twitterworks    1
Name: Company, Length: 114, dtype: int64

In [16]:
# Unique Companies
df["Company"].unique()

array(['Avamba', 'Fatz', 'Leexo', 'Rhyzio', 'Eadel', 'Oozz', 'Innotype',
       'Jayo', 'Riffpedia', 'Tazz', 'Browsezoom', 'Skinte', 'Twitterlist',
       'Shuffletag', 'Geba', 'Flipopia', 'Topicblab', 'Feedmix',
       'Skynoodle', 'Mita', 'Skinix', 'Aimbu', 'Babblestorm', 'Quinu',
       'Dynazzy', 'Tagtune', 'Kimia', 'Centizu', 'Muxo', 'Dynabox',
       'Skalith', 'Tagchat', 'Buzzdog', 'Zava', 'Ntags', 'Youspan',
       'Skyvu', 'Brightdog', 'Shufflebeat', 'Fivechat', 'Topiclounge',
       'Youopia', 'Dabjam', 'Kazio', 'Mycat', 'Meevee', 'Skibox',
       'Topiczoom', 'Realfire', 'Devshare', 'Devcast', 'Eare', 'Dabshots',
       'BlogXS', 'Rhycero', 'Meembee', 'Tagopia', 'Feedspan', 'Lajo',
       'Ainyx', 'Riffpath', 'Rhynyx', 'Feedfish', 'Kwinu', 'Skinder',
       'Jabbersphere', 'Tambee', 'Yadel', 'Flashpoint', 'Vipe',
       'Edgeclub', 'Jatri', 'Yozio', 'Oloo', 'Dazzlesphere', 'Yabox',
       'Miboo', 'Leenti', 'Quire', 'Meetz', 'Realcube', 'Fiveclub',
       'Buzzster', 'Camido

- By looking into the dataset, it is established that the nan values' corresponding value in date column might be representing the method/mode of booking the rooms.
- For the above measure, we will try to make a new column naming *mode of booking* or *booking method*. 

In [6]:
# Making a new column and filling the column with nan values
# Condition
con = df["Room number"].isna()
# Filling the column with date value otherwise with nan values
df["Mode_of_booking"] = np.where(con, df["Date"], np.nan)

In [7]:
df

Date       Company        Person Name  Room number Mode_of_booking
0    1-Jan-2022        Avamba  Anatole Ridehalgh       4008.0             NaN
1    1-Jan-2022          Fatz    Aldrich McKevin       2002.0             NaN
2    1-Jan-2022         Leexo    Stanley Hadrill       4012.0             NaN
3        Hotels           NaN                NaN          NaN          Hotels
4    1-Jan-2022        Rhyzio       Lyndell Tice       1006.0             NaN
..          ...           ...                ...          ...             ...
190   Cleartrip           NaN                NaN          NaN       Cleartrip
191  1-Jan-2022      Fivechat     Corabella Saye       4008.0             NaN
192  1-Jan-2022       Innojam  Leandra Potapczuk       5002.0             NaN
193  1-Jan-2022  Twitterworks    Valentia Ledson       1010.0             NaN
194      Hotels           NaN                NaN          NaN          Hotels

[195 rows x 5 columns]

In [8]:
# Filling the nan values with the text values in date column
df["Mode_of_booking"].bfill(inplace=True)

In [9]:
df

Date       Company        Person Name  Room number Mode_of_booking
0    1-Jan-2022        Avamba  Anatole Ridehalgh       4008.0          Hotels
1    1-Jan-2022          Fatz    Aldrich McKevin       2002.0          Hotels
2    1-Jan-2022         Leexo    Stanley Hadrill       4012.0          Hotels
3        Hotels           NaN                NaN          NaN          Hotels
4    1-Jan-2022        Rhyzio       Lyndell Tice       1006.0         Booking
..          ...           ...                ...          ...             ...
190   Cleartrip           NaN                NaN          NaN       Cleartrip
191  1-Jan-2022      Fivechat     Corabella Saye       4008.0          Hotels
192  1-Jan-2022       Innojam  Leandra Potapczuk       5002.0          Hotels
193  1-Jan-2022  Twitterworks    Valentia Ledson       1010.0          Hotels
194      Hotels           NaN                NaN          NaN          Hotels

[195 rows x 5 columns]

- As we have filled the nan values and mentioned the data in a sequence, we can now drop the nan values.

In [10]:
# Dropping the missing values

df.dropna(inplace=True)

In [11]:
df

Date       Company          Person Name  Room number  \
0    1-Jan-2022        Avamba    Anatole Ridehalgh       4008.0   
1    1-Jan-2022          Fatz      Aldrich McKevin       2002.0   
2    1-Jan-2022         Leexo      Stanley Hadrill       4012.0   
4    1-Jan-2022        Rhyzio         Lyndell Tice       1006.0   
5    1-Jan-2022         Eadel  Broderic Handscombe       3015.0   
..          ...           ...                  ...          ...   
186  1-Jan-2022        Tagpad       Stephani Lafee       1015.0   
187  1-Jan-2022        Meevee      Victoria Lavery       7002.0   
191  1-Jan-2022      Fivechat       Corabella Saye       4008.0   
192  1-Jan-2022       Innojam    Leandra Potapczuk       5002.0   
193  1-Jan-2022  Twitterworks      Valentia Ledson       1010.0   

    Mode_of_booking  
0            Hotels  
1            Hotels  
2            Hotels  
4           Booking  
5           Booking  
..              ...  
186         Expedia  
187         Expedia  
191          Hotels  
192          Hotels  
193          Hotels  

[134 rows x 5 columns]

In [13]:
# Checking for missing values

df.isnull().sum()

Date               0
Company            0
Person Name        0
Room number        0
Mode_of_booking    0
dtype: int64

- After dropping the missing values we are left with **134 rows/instances** and **5 columns/variables**. 